liverestore: Don't remove `--rm` containers on restart

vvoland opened the pull request:
- fixes: https://github.com/moby/moby/issues/46308

When live-restore is enabled, containers with autoremove enabled shouldn't be forcibly killed when engine restarts.

**- What I did**

**- How I did it**
Added extra check and integration test.

**- How to verify it**
TestLiveRestore

**- Description for the changelog**
```release-note
- live restore: Containers with auto remove (`docker run --rm`) are no longer forcibly removed on engine restart.
```


**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
liverestore: Don't remove `--rm` containers on restart

When live-restore is enabled, containers with autoremove enabled
shouldn't be forcibly killed when engine restarts.
They still should be removed if they exited while the engine was down
though.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on daemon/daemon.go:
```diff
@@ -494,7 +494,7 @@ func (daemon *Daemon) restore(cfg *configStore) error {
 				restartContainers[c] = make(chan struct{})
 				mapLock.Unlock()
-			} else if c.HostConfig != nil && c.HostConfig.AutoRemove {
+			} else if c.HostConfig != nil && c.HostConfig.AutoRemove && !cfg.LiveRestoreEnabled {
```
The original commit had a check if the container was still running (or paused); https://github.com/moby/moby/commit/3c2886d8a45d8e79b00ab413d91f1af871b58d0a#diff-6a5ec84e73a526676a7b1e1f5e730211f009679a8736b7e7cda6a111a986ab4cR203-R205

```go
if !c.IsRunning() && !c.IsPaused() {
	if daemon.configStore.AutoRestart && c.ShouldRestart() {
		mapLock.Lock()
		restartContainers[c] = make(chan struct{})
		mapLock.Unlock()
	} else if c.HostConfig != nil && c.HostConfig.AutoRemove {
		removeContainers[c.ID] = c
	}
}
```


I wonder if that's the bit that got lost, because there's multiple loops / checks (so perhaps we now fall through to code that shouldn't be run on "running" containers

vvoland replied:
Yes, we still should remove the container which exited while the engine was down.
Added the check and an integration test for this case.

cpuguy83 commented on integration/daemon/daemon_test.go:
```diff
@@ -381,6 +382,66 @@ func TestLiveRestore(t *testing.T) {
+		d.Stop(t)
+
+		finishContainer()
+		time.Sleep(time.Millisecond * 200)
```
This is racey.
We could improve this by having the container write a file on exit and then wait 200ms, though that's still racey.

Another option would be to use a fifo, have the container open the fifo as a writer and the test open the file as read.
WHen the container exits the write side will be closed and read side will get EOF.

vvoland replied:
Theoretically yes, in practice I think there's a very low probability that the container won't finish before the daemon is started.

With the FIFO I think there's still a race between closing the FIFO and the process being actually removed?

A simpler solution (and truly race-free I think?) would be to get the container process PID and just wait until it it's finished (`os.FindProcess` + `process.Wait()`). WDYT?

cpuguy83 replied:
> in practice I think there's a very low probability that the container won't finish before the daemon is started.

I've seen some stuff in our CI.
If it can go wrong it will.

> With the FIFO I think there's still a race between closing the FIFO and the process being actually removed?

Removed as in the shim seeing that it's down? Perhaps, but the fd will only be closed because the process exited.
Certainly some possible timing issues there in terms of what the shim would report, but I think that's the best we're going to get outside of making a separate request directly to the shim.

> A simpler solution (and truly race-free I think?) would be to get the container process PID and just wait until it it's finished (os.FindProcess + process.Wait()). WDYT?

I don't think you'll be able to wait for the process.
But... maybe in this case it will work since the test process would be the parent of dockerd->containerd->shim->process
Probably this has the same exact slight timing issue as above.
If you think this is simpler, that seems fine, though.

cpuguy83 replied:
Oh well, actually no.
This would use the main containerd process started with the test suite as I recall.

cpuguy83 replied:
If you want to wait on the pid, I think you'd need to first start an init container and then use that pid namespace for the test container.
This way the parent is the first container, as PID 1.

vvoland replied:
Ah, right. I forgot that we can't `waitpid` on a non-child process...
There's still the `kill -0` trick (and we even seem to use it in `pkg/process`).

I did a quick implementation. Let me know if that looks good to you.

vvoland committed:
integration/TestLiveRestore: Wait for process to exit

Replace `time.Sleep` with a poll that checks if process no longer exists
to avoid possible race condition.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

cpuguy83 reviewed (approved):
LGTM

akerouanton reviewed (approved):
LGTM. Thanks for taking care of this!

vvoland commented:
Thanks for reviewing!
